trunk/modules/python/dataset/reader.py

```diff
@@ -144,38 +205,76 @@ def makePILImageBandContiguous(img, log=None) :
         a = np.asarray(img.getdata(), dtype=t.config.floatX)
         return np.resize(statisticalNorm(a), (1, img.size[1], img.size[0]))
 
-def readPILImage(image, log=None) :
+def readPILImage(image, log=None, **kwargs) :
     '''This method should be used for all regular image formats from JPEG,
        PNG, TIFF, etc. A PIL error may originate from this method if the image
        format is unsupported.
     '''
     from PIL import Image
-    img = Image.open(image)
-    img.load() # because PIL can be lazy
-    return makePILImageBandContiguous(img)
+    from io import BytesIO
+
+    # PIL images can be a bunch of different types
+    # with different memory layouts that are nasty
+    # to recreate (putting them as numpy arrays
+    # disturbs their delicate nature)
+    #
+    # So for 'raw' input, just punt on interpreting
+    # them at all and just grab the raw file bytes.
+    # 
+    # When we need the 'processed' version, PIL
+    # is smart enough to decode these in memory
 
-def readImage(image, log=None) :
+    if type(image) == tuple:
+        # pass in buffer and type as tuple
+        # read raw bytes
+        img = Image.open(BytesIO(image[0].tobytes()))
+    else:
+        img = Image.open(image)
+        img.load() # because PIL can be lazy
+
+    raw = kwargs.get('raw', False)
+    if not raw:
+        out = makePILImageBandContiguous(img, log)
+    else:
+        if type(image) == str:
+            out = (np.fromfile(image, dtype=np.uint8),
+                   img.mode)
+        else:
+            # just hand back the input
+            out = image
+
+    return out
+
+def readImage(image, log=None, **kwargs) :
```

Comment:
The meat and potatoes of the hdf5 vs filesystem should be here. If the "image" is a str, call the appropriate open* method, however otherwise just remap it according to the shape and dtype, and return. After looking through here I do think it should be the responsibility of the individual open* methods to perform the atleastND() or the alternative would be to add a openImage() method which does this across all image formats.

This second path could be useful for creating the hdf5 filesystem file in dir2hdf5. It'll get you your data, non-remapped and into the channel format you need (without having redundant logic everywhere). My vote is now on the second path